### ISSUE TITLE
Add method __toString() to MangoPay\Libraries\Error

### ISSUE THREAD

**Issue by shakaran** [Mangopay/mangopay2-php-sdk#76]
The [MangoPay\Libraries\Error](https://github.com/Mangopay/mangopay2-php-sdk/blob/master/MangoPay/Libraries/Error.php) doesn't have a method __toString() so I cannot convert easy error messages to string when I catch a [ResponseException](https://github.com/Mangopay/mangopay2-php-sdk/blob/master/MangoPay/Libraries/ResponseException.php) since it is a stdClass object and the only solution is serialize or parse the object.

For example:

	use MangoPay\Libraries\ResponseException;
	use MangoPay\Libraries\Logs as MangoLog;
	use MangoPay\Libraries\Exception as MangoException;

	// my controller code
	
	// ...

	try 
	{
		$bankAccount = $mangopay->createBankAccount($user, $details);
	}
	catch(ResponseException $e) 
	{
		MangoLog::Debug('MangoPay\ResponseException Code', $e->GetCode());
		MangoLog::Debug('Message', $e->GetMessage());
		MangoLog::Debug('Details', $e->GetErrorDetails());
		
		$mangoErrorMessage = 'MangoPay[' . intval($e->GetCode()) . ']: ' . $e->GetMessage() . ' <br />' . serialize($e->GetErrorDetails()->Errors);
		redirect(route('someroute'))->with('error_message', $mangoErrorMessage);
	} 
	catch(MangoException $e) 
	{
		MangoLog::Debug('MangoPay\Exception Message', $e->getTraceAsString());
		
		$mangoErrorMessage = 'MangoPay[' . intval($e->getCode()) . ']: ' . $e->getTraceAsString();
		redirect(route('someroute'))->with('error_message', $mangoErrorMessage);
	}

As you see, I have to use serialize($e->GetErrorDetails()->Errors), because I didn't have a getErrors() or a way to __toString() in that class and Log::Debug is is mostly using a print_r() for this situations that is not good if you want store it in a variable and show as error message in a webapp or store it in a log file.

So actually using a contatenation in a string like:

    $myvar = 'Some message: ' . $e->GetErrorDetails()->Errors . ' oops'; 

Will give you a FatalError exception and breaks all the app.

This is probably a easy change and add that methods will not make a BC I think.

**Comment by hobailey** [comment]
Hi @shakaran, thanks for bringing this up. It's an interesting point, but I can't quite see how to update the SDK without creating a breaking change though but perhaps I've misunderstood. Could you perhaps submit a pull request offering a solution that responds to your need?

**Comment by shakaran** [comment]
@hobailey I did a small PR with the change that I try to explain you. I think that it will not introduce any BC at least running the tests is working fine.

**Comment by nclavaud** [comment]
> The MangoPay\Libraries\Error doesn't have a method __toString() so I cannot convert easy error messages to string when I catch a ResponseException since it is a stdClass object and the only solution is serialize or parse the object.

I would not give the responsibility of formatting error messages to this class. Formatting needs can be so specific: you might want text output, or json, or html, or whatever.

Delegating error formatting to an external method/class looks like a better option to me, and enforces single responsibility.

And actually, it's quite easy to iterate over the messages since the object has its properties public and you can use basic [object iteration](http://php.net/manual/en/language.oop5.iterations.php) or [cast this object into an array](http://php.net/manual/en/language.types.array.php#language.types.array.casting):
```php
$errors = $e->GetErrorDetails()->Errors;
$myvar = 'Some message: ' . TextErrorFormatter::format($errors) . ' oops';
$myvarHtml = '<strong>Errors</strong>' . HtmlErrorFormatter::format($errors);

class HtmlErrorFormatter
{
    public static function format($errors)
    {
        $formattedError = '<ul>';

        // object iteration
        foreach ($errors as $key => $message) {
            $formattedError .= sprintf('<li>%s: %s</li>', $key, $message);
        }

        $formattedError .= '</ul>';

        return $formattedError;
    }
}

class TextErrorFormatter
{
    public static function format($errors)
    {
        // cast to array
        return implode(' - ', (array) $errors);
    }
}
```

**Comment by shakaran** [comment]
@nclavaud very useful comment :+1: 

Probably the general users only wants a way to display easily the error messages, so I think that MangoPay just should give a default way to do it.

Even giving a default way, you can always pick the object as you show and rewrite the formatting error as you desire.

For example, in Log Debug they are giving us a default way to show the error log data:

https://github.com/Mangopay/mangopay2-php-sdk/blob/master/MangoPay/Libraries/Logs.php#L13

It can be rewriten if the user don't want a print_r and change for a json, or with html formatting, etc. The subject is give to the users a default way to display the data, it is up the users change the default formatting if they desire.

**Comment by hobailey** [comment]
Hey, coming back to this, I've just merged in @shakaran's #80 - let's see how that goes, but do come back to this thread if you any further ideas/inputs for this. Thanks to you both for your help :-)
